### DISCUSSION TITLE
I've added FTDI (MPSSE) adapter support

### DISCUSSION THREAD

**Discussion by nattgris** [pyocd/pyOCD#1701] [Show and tell]
Hi!

I have recently published [pyocd-mpsse](https://github.com/nattgris/pyocd-mpsse) to PyPi, adding support for FTDI based debug adapters (only for chips supporting the MPSSE mode, i.e. FT232H, FT2232 etc.).

Currently it's very crude with support only for SWD and performance is not great due to only using synchronous transfers and no queuing. Still, it seems to work fine for what have used it for.

I have hard-coded the USB VID:PID and the default I/O configuration to my own adapter that I use. I'm not sure if there is any existing support in pyocd for specifying USB properties in a generic way. If there's not, maybe the best option is to add session options for those as well. For now, if you want to use it, you need to change the [hardcoded](https://github.com/nattgris/pyocd-mpsse/blob/45a87703991fd56e6f8fc89ade1598211354ef85/pyocd_mpsse/mpsse_probe.py#L418) values to match your adapter. You also need to configure the options below to match, which lets the adapter driver know how the signals are mapped to the FTDI GPIOs.

```
ftdi.channel (int) FTDI channel.
fdti.gpio_dir_mask (int) GPIO initial direction bitmask.
ftdi.gpio_init_mask (int) GPIO initial output bitmask.
ftdi.jtag_en_mask (int) JTAG enable bitmask.
ftdi.srst_mask (int) SRST bitmask.
ftdi.swd_en_mask (int) SWD enable bitmask.
ftdi.swdio_oe_mask (int) SWDIO output enable pin bitmask.
```

The options are the same as used in OpenOCD, so you can take your values from the corresponding [config file](https://github.com/openocd-org/openocd/tree/master/tcl/interface/ftdi). Note however that not all variations possible in OpenOCD are supported yet, e.g. active low signals and tristate control, so some adapters may not be possible to support without further coding.

Feel free to open PRs adding the missing functionality, or just open an issue if you want to suggest some ideas. My own development on it will probably be rather sporadic.

**Comment by cederom** [comment]
THanks @nattgris :-) If you make implementation compatible with UrJTAG / OpenOCD / AvrDude we could reuse their drivers..? :-)

**Comment by nattgris** [comment]
The driver is naturally heavily based on the ftdi adapter driver in OpenOCD, since I'm the one who wrote that :smile: 

So I intend for configuration to be compatible (easily convertible), once the feature set is on par with OpenOCD.
